services/wallet/downloader.go

```diff
@@ -0,0 +1,159 @@
+package wallet
+
+import (
+	"context"
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+// TransferType type of the asset that was transferred.
+type TransferType string
+
+const (
+	ethTransfer   TransferType = "eth"
+	erc20Transfer TransferType = "erc20"
+
+	erc20TransferEventSignature = "Transfer(address,address,uint256)"
+)
+
+var (
+	one = big.NewInt(1)
+)
+
+// Transfer stores information about transfer.
+type Transfer struct {
+	Type        TransferType
+	Header      *types.Header
+	Transaction *types.Transaction
+	Receipt     *types.Receipt
+}
+
+// ETHTransferDownloader downloads regular eth transfers.
+type ETHTransferDownloader struct {
+	client  *ethclient.Client
+	address common.Address
+	signer  types.Signer
+}
+
+// GetTransfers checks is balance was changed between two blocks. And if so it downloads transfers
+// for that block.
+func (d *ETHTransferDownloader) GetTransfers(ctx context.Context, header *types.Header) (rst []Transfer, err error) {
+	// TODO(dshulyak) consider caching balance and reset it on reorg
+	num := new(big.Int).Sub(header.Number, one)
+	balance, err := d.client.BalanceAt(ctx, d.address, num)
+	if err != nil {
+		return nil, err
+	}
+	current, err := d.client.BalanceAt(ctx, d.address, header.Number)
+	if err != nil {
+		return nil, err
+	}
+	if current.Cmp(balance) == 0 {
+		return nil, nil
+	}
+	blk, err := d.client.BlockByHash(ctx, header.Hash())
+	if err != nil {
+		return nil, err
+	}
+	for _, tx := range blk.Transactions() {
+		if *tx.To() == d.address {
+			receipt, err := d.client.TransactionReceipt(ctx, tx.Hash())
+			if err != nil {
+				return nil, err
+			}
+			rst = append(rst, Transfer{Type: ethTransfer, Header: header, Transaction: tx, Receipt: receipt})
+			continue
+		}
+		from, err := types.Sender(d.signer, tx)
+		if err != nil {
+			return nil, err
+		}
+		// payload is empty for eth transfers
+		if from == d.address && len(tx.Data()) == 0 {
+			receipt, err := d.client.TransactionReceipt(ctx, tx.Hash())
+			if err != nil {
+				return nil, err
+			}
+			rst = append(rst, Transfer{Type: ethTransfer, Header: header, Transaction: tx, Receipt: receipt})
+			continue
+		}
+	}
+	if len(rst) == 0 {
+		return nil, errors.New("balance changed but no new transactions were found")
```

Comment:
Is this an error?

Reply:
yeah it is subtle but can happen. i query balance with block number and then query block body and because we are using infura we can hit different replica. it could be that we will hit replica A on first query and replica B on second. and block won't be the same there

but i realized that this is not the best way to test because gas spent will result in balance change as well